src/App.tsx

```diff
@@ -1,14 +1,103 @@
 /* eslint-disable jsx-a11y/label-has-associated-control */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, { useEffect, useRef, useState } from 'react';
 import { UserWarning } from './UserWarning';
-import { USER_ID } from './api/todos';
+import { getTodos, updateTodo, USER_ID } from './api/todos';
+import { Todo } from './types/Todo';
+import classNames from 'classnames';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[] | null>(null);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const [updatingTodoId, setUpdatingTodoId] = useState<number | null>(null);
+  const [allTodos, setAllTodos] = useState<Todo[] | null>(null);
+  const [activeFilter, setActivefilter] = useState<string>('all');
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+        setAllTodos(todosFromServer);
+        setErrorMessage('');
+        inputRef.current?.focus();
+      })
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+      });
+  }, []);
+
+  const toggleTodo = async (id: number) => {
+    setUpdatingTodoId(id);
+    const todoToUpdate = todos?.find(todo => todo.id === id);
+
+    if (!todoToUpdate) {
+      return;
+    }
+
+    try {
+      await updateTodo(id, { completed: !todoToUpdate.completed });
+      setTodos(prevTodos =>
+        (prevTodos as Todo[]).map(todo =>
+          todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+        ),
+      );
+      setAllTodos(prevTodos =>
+        (prevTodos as Todo[]).map(todo =>
+          todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+        ),
+      );
+    } catch {
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setUpdatingTodoId(null);
+    }
+  };
+
+  function handleFilter(filterType: string) {
```

Comment:
It is important to follow the same way of creating functions everywhere
```suggestion
  const handleFilter = (filterType: string) => {
```